File: java/leetcode_grind/Day571.java
package leetcode_grind;

import java.util.HashMap;

public class Day571 {
    // https://leetcode.com/problems/continuous-subarray-sum/description
    static class Solution1 {
        public boolean checkSubarraySum(int[] nums, int k) {
            int prefixMod = 0;
            HashMap<Integer, Integer> modSeen = new HashMap<>();
            modSeen.put(0, -1);

            for (int i = 0; i < nums.length; i++) {
                prefixMod = (prefixMod + nums[i]) % k;

                if (modSeen.containsKey(prefixMod)) {
                    if (i - modSeen.get(prefixMod) > 1) {
                        return true;
                    }
                } else {
                    modSeen.put(prefixMod, i);
                }
            }

            return false;
        }
    }
}

File: python/leetcode_grind/day_571.py
# https://leetcode.com/problems/continuous-subarray-sum/description
class Solution1:
    def checkSubarraySum(self, nums: List[int], k: int) -> bool:
        prefix_mod = 0
        mod_seen = {0: -1}
        for i in range(len(nums)):
            prefix_mod = (prefix_mod + nums[i]) % k
            if prefix_mod in mod_seen:
                if i - mod_seen[prefix_mod] > 1:
                    return True
            else:
                mod_seen[prefix_mod] = i
        return False

# https://leetcode.com/problems/subarray-sum-equals-k/description/
class Solution2:
    def subarraySum(self, nums: List[int], k: int) -> int:
        m = {0: 1}
        s = 0
        ans = 0
        for n in nums:
            s += n
            if s - k in m:
                ans += m[s-k]
            if not s in m:
                m[s] = 0
            m[s] += 1
        return ans

# https://leetcode.com/problems/minimum-number-of-operations-to-make-array-continuous/description/
class Solution3:
    def minOperations(self, nums: List[int]) -> int:
        n = len(nums)
        ans = n
        new_nums = sorted(set(nums))
        j = 0
        for i in range(len(new_nums)):
            while j < len(new_nums) and new_nums[j] < new_nums[i] + n:
                j += 1
            count = j - i
            ans = min(ans, n - count)
        return ans